# T05_Motor Trend Car Road Tests

|                |   |
:----------------|---|
| **Nombre**     | Jorge Oviedo Magaña   |
| **Fecha**      | 17/02/2026  |
| **Expediente** | 757048   |

# Regresión Lineal sobre Motor Trend Car Road Tests

## Dataset
Motor Trend Car Road Tests (mtcars) contiene información de
32 autos con variables como cilindros, peso, potencia, etc.

## Objetivo
Aprender a construir y evaluar modelos de regresión lineal
variando tres elementos:

- **Variable de salida** — `mpg` (consumo) o `qsec` (velocidad)
- **Tratamiento de predictores** — numéricos puros o con dummies
- **Modelo** — OLS simple o Ridge con regularización

## Flujo de trabajo
Todos los ejercicios siguen el mismo esquema:
```
Dataset → Split (40% train / 60% test) → Escalado → Modelo → R² Train
 → R² Test
 → Comparar
```

## Ejercicios

| Ejercicio | Salida  | Predictores     | Modelo       |
|-----------|---------|-----------------|--------------|
| 1.1       | `mpg`   | Numéricos       | OLS + Ridge  |
| 1.2       | `qsec`  | Numéricos       | OLS + Ridge  |
| 2.1       | `mpg`   | Numéricos + dummies | OLS     |
| 2.2       | `qsec`  | Numéricos + dummies | OLS     |
| 3.1       | `mpg`   | Comparación 1.1 vs 2.1 | —    |
| 3.2       | `qsec`  | Comparación 1.2 vs 2.2 | —    |

### **1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'.**

In [22]:
import pandas as pd

Motor = pd.read_excel("Motor Trend Car Road Tests.xlsx")

In [23]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import r2_score

Motor = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# ── EDA / Split ───────────────────────────────────────────
muestra = Motor.sample(frac=0.40, random_state=123)   # 40% train

Y = muestra["mpg"]
X = muestra.drop(["mpg", "model"], axis=1)

Xtest = Motor.drop(muestra.index).drop(["mpg", "model"], axis=1)
Ytest = Motor.drop(muestra.index)["mpg"]

# ── Escalado ──────────────────────────────────────────────
scaler = StandardScaler()
Xfinal      = scaler.fit_transform(X)       # fit solo en train
Xtest_final = scaler.transform(Xtest)       # solo transform en test

# ── 1. Regresión Lineal (OLS) ─────────────────────────────
modelo = LinearRegression()
modelo.fit(Xfinal, Y)

# R² train
Ypred_train = modelo.predict(Xfinal)
r2_train1 = r2_score(Y, Ypred_train)

# R² test
Ypred_test = modelo.predict(Xtest_final)
r2_test1 = r2_score(Ytest, Ypred_test)

print("=== OLS ===")
print(f"R² Train : {r2_train1:.4f}")
print(f"R² Test  : {r2_test1:.4f}")

# Betas con sus nombres
betas = pd.Series(modelo.coef_, index=X.columns)
print("\nBetas (coeficientes estandarizados):")
print(betas.sort_values())

# ── 2. Ridge con varios lambdas ───────────────────────────
lambdas = [0.01, 0.1, 1, 10, 50, 100, 500, 1000]

print("\n=== Ridge ===")
print(f"{'Lambda':>8}  {'R² Train':>10}  {'R² Test':>10}")

for lam in lambdas:
    ridge = Ridge(alpha=lam)
    ridge.fit(Xfinal, Y)
    rt  = r2_score(Y,     ridge.predict(Xfinal))
    rte = r2_score(Ytest, ridge.predict(Xtest_final))
    print(f"{lam:>8}  {rt:>10.4f}  {rte:>10.4f}")

=== OLS ===
R² Train : 0.9796
R² Test  : -1.1210

Betas (coeficientes estandarizados):
carb    -3.381685
disp    -3.023609
hp      -2.688186
am      -2.332915
qsec    -0.938253
wt       2.287884
drat     6.647521
vs       6.741944
gear    11.091457
cyl     17.646645
dtype: float64

=== Ridge ===
  Lambda    R² Train     R² Test
    0.01      0.9759      0.0525
     0.1      0.9632      0.5807
       1      0.9419      0.7404
      10      0.8793      0.7935
      50      0.7427      0.6523
     100      0.6005      0.5048
     500      0.2228      0.1248
    1000      0.1238      0.0266


## **1.2 Repite el ejercicio anterior usando 'qsec' como salida.**

In [24]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import r2_score

Motor = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# ── EDA / Split ───────────────────────────────────────────
muestra = Motor.sample(frac=0.40, random_state=123)   # 40% train

Y = muestra["qsec"]
X = muestra.drop(["qsec", "model"], axis=1)

Xtest = Motor.drop(muestra.index).drop(["qsec", "model"], axis=1)
Ytest = Motor.drop(muestra.index)["qsec"]

# ── Escalado ──────────────────────────────────────────────
scaler = StandardScaler()
Xfinal      = scaler.fit_transform(X)       # fit solo en train
Xtest_final = scaler.transform(Xtest)       # solo transform en test

# ── 1. Regresión Lineal (OLS) ─────────────────────────────
modelo = LinearRegression()
modelo.fit(Xfinal, Y)

# R² train
Ypred_train = modelo.predict(Xfinal)
r2_train2 = r2_score(Y, Ypred_train)

# R² test
Ypred_test = modelo.predict(Xtest_final)
r2_test2 = r2_score(Ytest, Ypred_test)

print("=== OLS ===")
print(f"R² Train : {r2_train2:.4f}")
print(f"R² Test  : {r2_test2:.4f}")

# Betas con sus nombres
betas = pd.Series(modelo.coef_, index=X.columns)
print("\nBetas (coeficientes estandarizados):")
print(betas.sort_values())

# ── 2. Ridge con varios lambdas ───────────────────────────
lambdas = [0.01, 0.1, 1, 10, 50, 100, 500, 1000]

print("\n=== Ridge ===")
print(f"{'Lambda':>8}  {'R² Train':>10}  {'R² Test':>10}")

for lam in lambdas:
    ridge = Ridge(alpha=lam)
    ridge.fit(Xfinal, Y)
    rt  = r2_score(Y,     ridge.predict(Xfinal))
    rte = r2_score(Ytest, ridge.predict(Xtest_final))
    print(f"{lam:>8}  {rt:>10.4f}  {rte:>10.4f}")

=== OLS ===
R² Train : 0.9585
R² Test  : -8.4729

Betas (coeficientes estandarizados):
disp   -4.072883
am     -2.688179
carb   -2.599902
mpg    -0.695778
vs      1.140538
wt      1.262623
hp      1.512581
gear    1.576697
drat    2.526791
cyl     3.455550
dtype: float64

=== Ridge ===
  Lambda    R² Train     R² Test
    0.01      0.9570     -8.1112
     0.1      0.9394     -3.5682
       1      0.8758      0.2616
      10      0.7077      0.8554
      50      0.3963      0.6717
     100      0.2633      0.4341
     500      0.0752     -0.0061
    1000      0.0401     -0.0978


### **2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.**

In [25]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

Motor = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# ── Dummies ───────────────────────────────────────────────
# drop_first=True evita la trampa de la variable dummy (multicolinealidad perfecta)
Motor = pd.get_dummies(Motor, columns=["cyl", "gear", "carb"], drop_first=True)

# ── EDA / Split ───────────────────────────────────────────
muestra = Motor.sample(frac=0.40, random_state=123)

Y = muestra["mpg"]
X = muestra.drop(["mpg", "model"], axis=1)

Xtest = Motor.drop(muestra.index).drop(["mpg", "model"], axis=1)
Ytest = Motor.drop(muestra.index)["mpg"]

# ── Escalado ──────────────────────────────────────────────
scaler = StandardScaler()
Xfinal      = scaler.fit_transform(X)
Xtest_final = scaler.transform(Xtest)

# ── Regresión OLS ─────────────────────────────────────────
modelo = LinearRegression()
modelo.fit(Xfinal, Y)

# ── R² Train ──────────────────────────────────────────────
r2_train3 = r2_score(Y, modelo.predict(Xfinal))

# ── R² Test ───────────────────────────────────────────────
r2_test3 = r2_score(Ytest, modelo.predict(Xtest_final))

print(f"R² Train : {r2_train3:.4f}")
print(f"R² Test  : {r2_test3:.4f}")

# ── Betas ─────────────────────────────────────────────────
betas = pd.Series(modelo.coef_, index=X.columns)
print("\nBetas (coeficientes estandarizados):")
print(betas.sort_values())

R² Train : 1.0000
R² Test  : -25294.9440

Betas (coeficientes estandarizados):
disp     -869.046836
am       -561.860633
qsec     -344.622910
carb_4   -306.885759
gear_5   -197.104324
vs       -193.998998
carb_3   -168.876933
cyl_8     -93.605338
wt        -85.946168
carb_6      0.000000
carb_8      0.000000
gear_4     46.696414
carb_2     97.849030
drat      138.882673
cyl_6     263.587483
hp        662.291850
dtype: float64


### **2.2 Repite el ejercicio anterior usando 'qsec' como salida.**

In [26]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

Motor = pd.read_excel("Motor Trend Car Road Tests.xlsx")

# ── Dummies ───────────────────────────────────────────────
Motor = pd.get_dummies(Motor, columns=["cyl", "gear", "carb"], drop_first=True)

# ── EDA / Split ───────────────────────────────────────────
muestra = Motor.sample(frac=0.40, random_state=123)

Y = muestra["qsec"]                                    # ← cambio
X = muestra.drop(["qsec", "model"], axis=1)            # ← cambio

Xtest = Motor.drop(muestra.index).drop(["qsec", "model"], axis=1)
Ytest = Motor.drop(muestra.index)["qsec"]              # ← cambio

# ── Escalado ──────────────────────────────────────────────
scaler = StandardScaler()
Xfinal      = scaler.fit_transform(X)
Xtest_final = scaler.transform(Xtest)

# ── Regresión OLS ─────────────────────────────────────────
modelo = LinearRegression()
modelo.fit(Xfinal, Y)

# ── R² Train ──────────────────────────────────────────────
r2_train4 = r2_score(Y, modelo.predict(Xfinal))

# ── R² Test ───────────────────────────────────────────────
r2_test4 = r2_score(Ytest, modelo.predict(Xtest_final))

print(f"R² Train : {r2_train4:.4f}")
print(f"R² Test  : {r2_test4:.4f}")

# ── Betas ─────────────────────────────────────────────────
betas = pd.Series(modelo.coef_, index=X.columns)
print("\nBetas (coeficientes estandarizados):")
print(betas.sort_values())

R² Train : 1.0000
R² Test  : -10.9301

Betas (coeficientes estandarizados):
disp     -4.795982
am       -3.100723
carb_4   -1.693601
gear_5   -1.087754
vs       -1.070616
carb_3   -0.931976
cyl_8    -0.516577
wt       -0.474308
mpg      -0.028806
carb_6    0.000000
carb_8    0.000000
gear_4    0.257702
carb_2    0.539996
drat      0.766448
cyl_6     1.454652
hp        3.654970
dtype: float64


### **Comparar R^2 de los ejercicios 1.1 y 2.1**

In [27]:
print(f"R² Train 1.1 (numéricos): {r2_train1:.4f}")
print(f"R² Train 2.1 (dummies)  : {r2_train2:.4f}")

print(f"R² Test 1.1 (numéricos): {r2_test1:.4f}")
print(f"R² Test 2.1 (dummies)  : {r2_test2:.4f}")

R² Train 1.1 (numéricos): 0.9796
R² Train 2.1 (dummies)  : 0.9585
R² Test 1.1 (numéricos): -1.1210
R² Test 2.1 (dummies)  : -8.4729


### Train
Ambos modelos ajustan bien en entrenamiento (~0.96–0.98),
lo cual es esperado con tan pocos datos (~13 observaciones).

### Test
Ambos R² Test son negativos, lo que significa que los modelos
son **peores que predecir la media** de `mpg`. Esto es sobreajuste
clásico: demasiados parámetros para muy pocas observaciones.

### ¿Por qué 2.1 es peor?
Las dummies sobre `cyl`, `gear` y `carb` añaden columnas extra,
aumentando aún más la relación parámetros/datos y agravando el sobreajuste.

## Conclusión
Con un dataset tan pequeño como mtcars (32 filas, 40% ≈ 13 train),
tratar las variables como **numéricas es preferible** a crear dummies.
Las dummies son útiles cuando hay suficientes datos para absorber
el incremento de dimensionalidad.

### **Comparar R^2 de los ejercicios 1.2 y 2.2**

In [29]:

print(f"R² Train 1.2 (numéricos): {r2_train3:.4f}")
print(f"R² Train 2.2 (dummies)  : {r2_train4:.4f}")

print(f"R² Test 1.2 (numéricos): {r2_test3:.4f}")
print(f"R² Test 2.2 (dummies)  : {r2_test4:.4f}")

R² Train 1.2 (numéricos): 1.0000
R² Train 2.2 (dummies)  : 1.0000
R² Test 1.2 (numéricos): -25294.9440
R² Test 2.2 (dummies)  : -10.9301


### Train
Ambos modelos memorizan perfectamente el train (R² = 1.0),
lo cual no es un logro sino una señal de sobreajuste total.

### Test
Aquí ocurre algo interesante e **inverso** a lo que pasó en 1.1 vs 2.1:

- En 1.1 vs 2.1 las dummies **empeoraban** el test (–1.12 → –8.47)
- En 1.2 vs 2.2 las dummies **mejoran** el test (–25,294 → –10.93)

### ¿Por qué las dummies ayudan con `qsec`?
El modelo numérico 1.2 trata `cyl`, `gear` y `carb` como
valores continuos, lo que genera coeficientes extremadamente
inestables con `qsec` como salida. Al convertirlas a dummies
en 2.2 se rompe esa linealidad forzada y los coeficientes
se estabilizan parcialmente, reduciendo el colapso en test.

## Conclusión
Aunque ninguno de los dos modelos generaliza bien, las dummies
son **claramente menos malas** con `qsec` como salida. El problema
de fondo sigue siendo el mismo: dataset demasiado pequeño para
OLS sin regularización.